File: app/src/androidTest/java/pl/srw/billcalculator/testutils/HistoryGenerator.java
package pl.srw.billcalculator.testutils;

import android.util.Log;

import org.threeten.bp.LocalDate;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

import pl.srw.billcalculator.db.PgeG11Bill;
import pl.srw.billcalculator.db.PgePrices;
import pl.srw.billcalculator.db.dao.DaoSession;
import pl.srw.billcalculator.db.dao.PgeG11BillDao;
import pl.srw.billcalculator.persistence.Database;
import pl.srw.billcalculator.util.Dates;

@Singleton
public final class HistoryGenerator {

    private final pl.srw.billcalculator.settings.prices.PgePrices prefsPrices;

    @Inject
    public HistoryGenerator(pl.srw.billcalculator.settings.prices.PgePrices pgePrices) {
        prefsPrices = pgePrices;
    }

    public void generatePgeG11Bills(final int count) {
        final PgeG11BillDao dao = Database.getSession().getPgeG11BillDao();
        final PgePrices pgePrices = prefsPrices.convertToDb();
        Log.d("", "insert " + count);
        Database.getSession().insert(pgePrices);
        List<PgeG11Bill> pgeBills = new ArrayList<>(count);
        for (int i = 1; i <= count; i++) {
            final int day = (i%335 == 0) ? 1 : i%335 ;
            final Date fromDate = Dates.toDate(LocalDate.ofYearDay(2014, day));
            final Date toDate = Dates.toDate(LocalDate.ofYearDay(2014, day + 30));
            pgeBills.add(new PgeG11Bill(null, i, i+10, fromDate, toDate, i*11.11, pgePrices.getId()));
        }

        dao.insertInTx(pgeBills);
    }

    public void clear() {
Log.d("", "clean");
        final DaoSession session = Database.getSession();
        //bills
        session.getPgnigBillDao().deleteAll();
        session.getPgeG11BillDao().deleteAll();
        session.getPgeG12BillDao().deleteAll();
        session.getTauronG11BillDao().deleteAll();
        session.getTauronG12BillDao().deleteAll();
        //prices
        session.getPgnigPricesDao().deleteAll();
        session.getPgePricesDao().deleteAll();
        session.getTauronPricesDao().deleteAll();
    }

    public void generatePgeG11Bill(int readingTo) {
        final PgeG11BillDao dao = Database.getSession().getPgeG11BillDao();
        final PgePrices pgePrices = prefsPrices.convertToDb();
System.out.println("insert " + pgePrices);
        Database.getSession().insert(pgePrices);
        final int day = (readingTo%335 == 0) ? 1 : readingTo%335 ;
        final Date fromDate = Dates.toDate(LocalDate.ofYearDay(2014, day));
        final Date toDate = Dates.toDate(LocalDate.ofYearDay(2014, day + 30));

        final int readingFrom = (readingTo - 10) < 0 ? 0 : readingTo-10;
        dao.insert(new PgeG11Bill(null, readingFrom, readingTo, fromDate, toDate, readingTo * 11.11, pgePrices.getId()));
    }
}

File: app/src/debug/java/pl/srw/billcalculator/history/HistoryGenerator.kt
package pl.srw.billcalculator.history

import org.threeten.bp.LocalDate
import pl.srw.billcalculator.data.bill.HistoryRepo
import pl.srw.billcalculator.db.PgeG11Bill
import pl.srw.billcalculator.db.PgeG12Bill
import pl.srw.billcalculator.persistence.Database
import pl.srw.billcalculator.settings.prices.PgePrices
import pl.srw.billcalculator.util.Dates
import timber.log.Timber
import javax.inject.Inject
import javax.inject.Singleton

@Suppress("MagicNumber")
@Singleton
class HistoryGenerator @Inject constructor(
    private val prices: PgePrices,
    private val historyRepo: HistoryRepo
) {

    private val currentYear = LocalDate.now().year

    fun clear() {
        Timber.d("Clear Database")
        val session = Database.getSession()
        //bills
        session.pgnigBillDao.deleteAll()
        session.pgeG11BillDao.deleteAll()
        session.pgeG12BillDao.deleteAll()
        session.tauronG11BillDao.deleteAll()
        session.tauronG12BillDao.deleteAll()
        //prices
        session.pgnigPricesDao.deleteAll()
        session.pgePricesDao.deleteAll()
        session.tauronPricesDao.deleteAll()

        historyRepo.deleteBillsWithPrices(emptyList()) // FIXME: Workaround to load latest state
    }

    fun generatePgeG11Bill(readingTo: Int) {
        Timber.d("Create PGE G11 bill with readingTo=$readingTo")
        insertBill(readingTo, insertPrices().id)
    }

    fun generatePgeG12Bill(readingDayTo: Int, readingNightTo: Int) {
        Timber.d("Create PGE G12 bill with readingTo=$readingDayTo")
        insertBill(readingDayTo, readingNightTo, insertPrices().id)
    }

    fun generatePgeG11Bills(count: Int) {
        Timber.d("Generate $count PGE G11 bills")
        for (i in 1..count) {
            insertBill(i + 10, insertPrices().id)
        }
    }

    private fun insertBill(readingTo: Int, id: Long) {
        val day = if (readingTo % 335 == 0) 1 else readingTo % 335
        val date = LocalDate.ofYearDay(currentYear, day)
        val fromDate = Dates.toDate(date)
        val toDate = Dates.toDate(date.plusDays(30))

        val readingFrom = if (readingTo - 10 < 0) 0 else readingTo - 10
        val bill = PgeG11Bill(null, readingFrom, readingTo, fromDate, toDate, readingTo * 11.11, id)
        historyRepo.insert(bill)
    }

    private fun insertBill(readingDayTo: Int, readingNightTo: Int, id: Long) {
        val day = if (readingDayTo % 335 == 0) 1 else readingDayTo % 335
        val date = LocalDate.ofYearDay(currentYear, day)
        val fromDate = Dates.toDate(date)
        val toDate = Dates.toDate(date.plusDays(30))

        val readingDayFrom = if (readingDayTo - 10 < 0) 0 else readingDayTo - 10
        val readingNightFrom = if (readingNightTo - 10 < 0) 0 else readingNightTo - 10
        val bill = PgeG12Bill(
            null, readingDayFrom, readingDayTo, readingNightFrom, readingNightTo,
            fromDate, toDate, readingDayTo * 11.11, id
        )
        historyRepo.insert(bill)
    }

    private fun insertPrices(): pl.srw.billcalculator.db.PgePrices {
        val prices = prices.convertToDb()
        historyRepo.insert(prices)
        return prices
    }
}
